extends Node2D


enum Pol√≠ticasDeReposicionamento {
	NENHUM,
	ALVO,
	TODOS
}

enum Alvos {
	DONUT,
	HAMB√öRGUER,
	PIZZA,
	OVO_FRITO,
	MA√á√É,
	UVA,
	SORVETE,
	CUPCAKE,
	BR√ìCOLIS,
	PICOL√â
}

enum Velocidades {
	EST√ÅTICA,
	LENTA,
	M√âDIA,
	R√ÅPIDA
}

@onready var escala: float = get_viewport_rect().size.x / 1152

# Par√¢metros do sistema
var m√∫sica_desligada: bool = false;
var sons_mutados: bool = false;

# Par√¢metros do jogo
var velocidade: int = Velocidades.M√âDIA
var alvos_no_jogo: Array = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
var pol√≠tica_de_reposicionamento: int = Pol√≠ticasDeReposicionamento.TODOS
var alvo_atual: int = 0
var pontos_real: int = 0
var pontos_display: int = 0
var vidas: int = 10
var suporte: int = 0
var dura√ß√£o: float = 120.0
var mostrar_barra_de_tempo: bool = true
var animar: bool = true

# Vari√°veis para o log
var log_data: Array = []  # Armazena os dados do log
var id_sess√£o: String = ""  # ID √∫nico para a sess√£o
var data_sess√£o: String = ""  # Data da sess√£o
var tempo_resposta: float = 0.0
var timestamp_atual_sess√£o: float = 0.0  # Dura√ß√£o da sess√£o em segundos
var profissional_respons√°vel: String = "Profissional Padr√£o"  # Nome do profissional respons√°vel
var nome_jogo: String = "Toque Certo"  # Nome do jogo
var id_profissional: String = "ID_Padr√£o_Profissional"
var caminho: String = OS.get_system_dir(OS.SYSTEM_DIR_DOWNLOADS) + "/game_logs.csv"

# Called when the node enters the scene tree for the first time.
func _ready() -> void:
	pass


# Chamado a cada frame
func _process(delta: float) -> void:
	# Atualizar a dura√ß√£o da sess√£o
	timestamp_atual_sess√£o += delta
	tempo_resposta += delta

# Fun√ß√£o para gerar um ID √∫nico para a sess√£o
func gerar_id_√∫nico() -> String:
	return str(randi()) + str(Time.get_ticks_msec())  # Combina um n√∫mero aleat√≥rio com o tempo atual em milissegundos


# Fun√ß√£o para obter a data e hora atuais
func obter_data_hora_atual() -> String:
	var data_hora: String = Time.get_datetime_string_from_system()  # Formato: "YYYY-MM-DDTHH:MM:SS"
	return data_hora


func iniciar_m√∫sica() -> void:
	$"M√∫sicaDeFundo".play()


func parar_m√∫sica() -> void:
	$"M√∫sicaDeFundo".stop()


func iniciar_jogo(n√∫mero_de_alvos: int, tempo_de_dura√ß√£o: float, pol√≠tica_de_reposicionamento_do_jogo: int, velocidade_dos_alvos: int, id_prof: String = "", mostrar_tempo: bool = true, n√∫mero_de_vidas: int = 2147483647, animar_√≠cones: bool = true) -> void:
	alvos_no_jogo = Alvos.values()
	alvos_no_jogo.shuffle()
	alvos_no_jogo = alvos_no_jogo.slice(0, n√∫mero_de_alvos)
	
	dura√ß√£o = tempo_de_dura√ß√£o
	pol√≠tica_de_reposicionamento = pol√≠tica_de_reposicionamento_do_jogo
	velocidade = velocidade_dos_alvos
	alvo_atual = alvos_no_jogo.pick_random()
	id_profissional = id_prof
	mostrar_barra_de_tempo = mostrar_tempo
	vidas = n√∫mero_de_vidas
	animar = animar_√≠cones

	pontos_real = 0
	pontos_display = 0
	suporte = 0
	timestamp_atual_sess√£o = 0.0
	tempo_resposta = 0.0
	log_data = []

	# Gerar ID √∫nico e data da sess√£o
	id_sess√£o = gerar_id_√∫nico()
	data_sess√£o = obter_data_hora_atual()
	print("ID da Sess√£o: ", id_sess√£o)
	print("Data da Sess√£o: ", data_sess√£o)
	
	get_tree().change_scene_to_file("res://scenes/jogo_principal/jogo_principal.tscn")


# Fun√ß√£o de clique para verificar acerto
func clique(alvos: Array[int]) -> bool:
	# Se m√∫ltiplos alvos forem clicados em simult√¢neo devido √† propaga√ß√£o do clique, o certo se sobrep√µe aos errados
	if alvo_atual in alvos:
		pontos_real += 1
		pontos_display += 1
		
		var anterior: int = alvo_atual
		
		# O alvo novo sempre ser√° diferente do anterior
		while alvo_atual == anterior and alvos_no_jogo.size() > 1:
			alvo_atual = alvos_no_jogo.pick_random()

		# Adiciona os dados ao log
		log_data.append([id_sess√£o, id_profissional, data_sess√£o, timestamp_atual_sess√£o, nome_jogo, tempo_resposta, pontos_real, alvo_atual, true, suporte, Velocidades.find_key(velocidade), vidas]) # Exemplo de log com acerto
		salvar_logs_csv()
		
		tempo_resposta = 0.0
		
		return true
	else:
		if pontos_display >= 1:
			pontos_display -= 1
		
		pontos_real -= 1
		
		if vidas != 2147483647:
			vidas -= 1
		
		log_data.append([id_sess√£o, id_profissional, data_sess√£o, timestamp_atual_sess√£o, nome_jogo, tempo_resposta, pontos_real, alvo_atual, false, suporte, Velocidades.find_key(velocidade), vidas])  # Exemplo de log com erro
		salvar_logs_csv()
		
		if vidas == 0:
			finalizar_sess√£o()
		
		return false


func mudan√ßa_no_suporte() -> void:
	# Adiciona uma linha indicando o novo suporte
	var file: FileAccess = FileAccess.open(caminho, FileAccess.READ_WRITE)
	
	file.seek_end()  # Vai para o final do arquivo
	file.store_line("%s,%s,%s,%f,%s,%s,%d,%d,%s,%s,%s,%d" % [id_sess√£o, id_profissional, data_sess√£o, timestamp_atual_sess√£o, nome_jogo, "MUDAN√áA NO SUPORTE", pontos_real, alvo_atual, "N/A", suporte, Velocidades.find_key(velocidade), vidas])
	
	file.close()


# Fun√ß√£o para salvar os logs em um arquivo CSV
func salvar_logs_csv(limpar: bool = true) -> void:
	var file: FileAccess
	
	if !FileAccess.file_exists(caminho):
		# Se o arquivo n√£o existir, cria um novo e adiciona o cabe√ßalho
		file = FileAccess.open(caminho, FileAccess.WRITE)
		if file == null:
			print("‚ùå Erro ao criar o arquivo de log!")
			return
		
		# Escreve o cabe√ßalho apenas uma vez
		file.store_line("ID_Sessao,ID_Profissional,Data_Sessao,Duracao_Sessao,Nome_Jogo,Tempo_Resposta,Pontos,Alvo_Atual,Acerto,Suporte,Velocidade,Vidas")
		
		file.close()
	
	file = FileAccess.open(caminho, FileAccess.READ_WRITE)

	# Vai para o final do arquivo
	file.seek_end()

	# Armazena os logs
	for log_entry in log_data:
		file.store_line("%s,%s,%s,%f,%s,%f,%d,%s,%s,%s,%s,%d" % [
			log_entry[0], log_entry[1], log_entry[2], log_entry[3], log_entry[4],
			log_entry[5], log_entry[6], str(log_entry[7]), str(log_entry[8]), str(log_entry[9]),
			log_entry[10], log_entry[11]
		])
		
	if limpar:
		log_data = []

	file.close()
	print("‚úÖ Logs salvos em: ", file.get_path_absolute())


# Fun√ß√£o para finalizar a sess√£o e registrar o t√©rmino
func finalizar_sess√£o(mudar_para_cena_final: bool = true) -> void:
	# Adiciona uma linha indicando o t√©rmino da sess√£o
	var file: FileAccess = FileAccess.open(caminho, FileAccess.READ_WRITE)
	
	file.seek_end()  # Vai para o final do arquivo
	file.store_line("%s,%s,%s,%f,%s,%s,%d,%s,%s,%d,%s,%d" % [id_sess√£o, id_profissional, data_sess√£o, timestamp_atual_sess√£o, nome_jogo, "FIM DA SESS√ÉO", pontos_real, "N/A", "N/A", suporte, Velocidades.find_key(velocidade), vidas])
	
	file.close()
	print("‚úÖ T√©rmino da sess√£o registrado no log.")
	
	if mudar_para_cena_final:
		get_tree().change_scene_to_file("res://UI/tela_final.tscn")


# Fun√ß√£o para verificar se o arquivo foi salvo
func verificar_logs() -> void:
	if FileAccess.file_exists(caminho):
		print("üìÇ Arquivo encontrado!")
		
		var file: FileAccess = FileAccess.open(caminho, FileAccess.READ)
		while not file.eof_reached():
			print(file.get_line())  # Exibe cada linha do CSV no console
		file.close()
	else:
		print("‚ùå Arquivo N√ÉO encontrado!")


# Fun√ß√£o para sair do jogo
func sair_do_jogo() -> void:
	print("Saindo do jogo...")  # Exibe uma mensagem de sa√≠da
	get_tree().quit()  # Encerra o jogo
